Pal/src/host/Linux-SGX/protected_files.c

```diff
@@ -0,0 +1,766 @@
+
+    if (PF_FAILURE(status)) {
+        DEBUG_PF("calculating header MAC failed: %d\n", status);
+        goto out;
+    }
+
+    status = PF_STATUS_MAC_MISMATCH;
+    if (memcmp(tag, hdr->header_mac, PF_MAC_SIZE) != 0) {
+        DEBUG_PF("MAC mismatch: ");
+        HEXDUMP(tag);
+        __DEBUG_PF(" vs expected ");
+        __HEXDUMPNL(hdr->header_mac, PF_MAC_SIZE);
+        goto out;
+    }
+
+    status = PF_STATUS_SUCCESS;
+
+out:
+    if (pf->header && PF_FAILURE(status)) {
+        if (PF_FAILURE(cb_unmap(pf->header, PF_HEADER_SIZE)))
+            DEBUG_PF("(failure path) header unmap failed\n");
+        pf->header = NULL;
+    }
+
+    return status;
+}
+
+// (Internal) Create header for a new PF (erases any data in the file)
+static pf_status_t create_header(pf_context_t* pf, const char* prefix, const char* file_name) {
+    pf_status_t status;
+    bool trailing_slash = prefix[strlen(prefix) - 1] == '/';
+
+    status = cb_truncate(pf->handle, PF_HEADER_SIZE);
+    if (PF_FAILURE(status))
+        goto out;
+
+    // Map the header
+    status = cb_map(pf->handle, pf->mode, 0, PF_HEADER_SIZE, (void**)&pf->header);
+    if (PF_FAILURE(status))
+        goto out;
+
+    pf_header_t* hdr = pf->header;
+    memset(hdr, 0, PF_HEADER_SIZE);
+    hdr->version = PF_FORMAT_VERSION;
+
+    // Generate IV for the header
+    status = cb_crypto_random(hdr->header_iv, PF_IV_SIZE);
+    if (PF_FAILURE(status))
+        goto out;
+
+    // TODO: multiple allowed paths support
+    hdr->allowed_paths_size = strlen(file_name)
+        + 1 // slash
+        + strlen(prefix)
+        + 1; // path NULL-terminator
+
+    // Trailing slash will be stripped
+    if (trailing_slash)
+        hdr->allowed_paths_size--;
+
+    status = PF_STATUS_PATH_TOO_LONG;
+    if (hdr->allowed_paths_size > PF_HEADER_ALLOWED_PATHS_SIZE)
+        goto out;
+
+    DEBUG_PF("allowed_paths_size: %u\n", hdr->allowed_paths_size);
+
+    // TODO: multiple allowed paths support
+    if (trailing_slash) {
+        snprintf(hdr->allowed_paths, hdr->allowed_paths_size, "%s%s", prefix, file_name);
+    } else {
+        snprintf(hdr->allowed_paths, hdr->allowed_paths_size, "%s/%s", prefix, file_name);
+    }
+
+    status = PF_STATUS_SUCCESS;
+out:
+    if (PF_FAILURE(status)) {
+        if (pf->header) {
+            if (PF_FAILURE(cb_unmap(pf->header, PF_HEADER_SIZE))) {
+                DEBUG_PF("(failure path) header unmap failed\n");
+            }
+            pf->header = NULL;
+        }
+    }
+
+    return status;
+}
+
+// (Internal) Update header MAC for a writable PF and set underlying file size
+static pf_status_t update_header(pf_context_t* pf, size_t data_size) {
+    pf_status_t status;
+    pf_header_t* hdr = pf->header;
+
+    DEBUG_PF("pf %p, data size %lu->%lu\n", pf, hdr->data_size, data_size);
+
+    hdr->data_size = data_size;
+
+    // Calculate header MAC
+    status = cb_crypto_aes_gcm_encrypt(pf->key, PF_WRAP_KEY_SIZE, hdr->header_iv, PF_IV_SIZE,
```

Comment:
Issue found by Intel colleague->Michael Steiner. When updating header, new IV needs to be generated, to prevent IV re-use in header.